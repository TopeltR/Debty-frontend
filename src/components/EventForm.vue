<template>
    <div>
        <navbar></navbar>
        <background>
            <b-row>
                <b-col sm='12' md='6' offset-md='3' class='TM30px'>
                    <b-row>
                        <h2>Create Event</h2>
                        <font-awesome-icon id="info" icon='info-circle' class="ml-1"></font-awesome-icon>
                        <b-tooltip target="info"
                                   title="Create an Event by giving it an informative title and contacts, then add your contacts to it."
                                   placement="bottom"></b-tooltip>
                    </b-row>
                    <b-row>
                        <form @submit.prevent='' class='wide mt-3'>
                            <div class='form-group'>
                                <label for='title'>Title:</label>
                                <input type='text' class='form-control' v-model='title' id='title' required=''
                                       placeholder='Enter event title'>
                            </div>
                            <div class='form-group'>
                                <label for='description'>Description:</label>
                                <textarea id='description' class='form-control' v-model='description' type='text'
                                          placeholder='Enter event description' required=''></textarea>
                            </div>
                            <div class='form-group'>
                                <p v-if='people.length > 0'>People:</p>
                                <ul>
                                    <li v-for='person in people'>
                                        {{ getFullName(person) }}
                                    </li>
                                </ul>
                                <label for='user'>Add people:</label>
                                <b-row>
                                    <b-col cols='9' class='PR0px'>
                                        <autocomplete id='user' v-model='user' :placeholder='"Name"' :field='field'
                                                      :items='allPeople'
                                                      :key-extractor='getFullName'></autocomplete>
                                    </b-col>
                                    <b-col cols='3'>
                                        <button type='button' v-on:click='addPerson'
                                                class='btn btn-primary wide'>
                                            Add
                                        </button>
                                    </b-col>
                                </b-row>
                            </div>
                            <b-row class='mt-4'>
                                <b-col v-for="button in buttons" :cols="button.width" :offset="button.offset">
                                    <b-btn :variant='button.variant'
                                           v-on:click='button.handler'
                                           class='wide'>
                                        {{button.name}}
                                    </b-btn>
                                </b-col>
                            </b-row>
                        </form>
                    </b-row>
                </b-col>
            </b-row>
        </background>
    </div>
</template>

<script>
    import User from '../entities/User';
    import router from '../router.ts';
    import Background from '@/components/Background';
    import Navbar from '@/components/Navbar';
    import Autocomplete from '@/components/Autocomplete';
    import userStore from '@/stores/UserStore';

    export default {
        name: 'EventForm',
        components: {Background, Navbar, Autocomplete},
        props: {
            eventId: {
                type: Number,
                default: undefined,
            },
            buttons: {
                type: Array,
                default: () => [{
                    width: 12,
                    offset: 0,
                    name: "Create event",
                    handler: (eventForm, userStore) => {
                        if (eventForm.title && eventForm.description) {
                            userStore.getUser().then((user) => {
                                eventForm.people.push(user);
                                eventForm.$http.post('/events', {
                                    title: eventForm.title,
                                    people: eventForm.people,
                                    description: eventForm.description,
                                    owner: user,
                                }).then((result) => {
                                    router.push('/events/' + result.data.id);
                                });
                            });
                        }
                    },
                    variant: 'primary',
                }],
            },
        },
        data: () => ({
            title: '',
            description: '',
            bills: [],
            user: {},
            allPeople: [],
            people: [],
            field: {value: ''},
        }),
        mounted() {
            if (this.eventId) {
                this.$http.get('/events/' + this.eventId).then((data) => {
                    data = data.data;
                    this.title = data.title;
                    this.description = data.description;
                    this.people = data.people;
                    this.bills = data.bills;
                });
            }

            for (let i = 0; i < this.buttons.length; i++) {
                let button = this.buttons[i];
                let handler = button.handler;
                button.handler = () => {
                    handler(this, userStore);
                };
            }
            userStore.getUser().then((user) => {
                this.$http.get('/contact/id/' + user.id).then((data) => {
                    this.allPeople = data.data.filter((u) => u.email !== user.email);
                });
            });
        },
        methods: {
            getFullName(user) {
                return user.firstName + ' ' + user.lastName;
            },
            addPerson() {
                userStore.getUser().then((user) => {
                    if (this.user && this.user.firstName && !this.people.map((user) => user.email).includes(this.user.email)
                        && this.user.email !== user.email) {
                        this.people.push(this.user);
                        this.allPeople = this.allPeople.filter((person) => !this.people.includes(person));
                        this.field.value = '';
                    }
                });
            },
        },
    }
    ;
</script>

<style scoped>
    #user {
        margin-bottom: 5px;
    }

    #description {
        height: 100px;
    }

    .wide {
        width: 100%;
    }

    .PR0px {
        padding-right: 0;
    }

    .TM30px {
        margin-top: 30px;
    }

</style>
