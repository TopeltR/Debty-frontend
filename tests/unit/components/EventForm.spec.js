import EventForm from '@/components/EventForm';
import userStore from '@/stores/UserStore'
import {mount} from "@vue/test-utils";
import Vue from "vue";
import BootstrapVue from 'bootstrap-vue';

describe('EventForm', () => {
    Vue.use(BootstrapVue);
    console.warn = jest.fn();
    console.error = jest.fn();

    describe('mounted', () => {
        it('should get event info if eventId is specified', async () => {
            const get = jest.fn(() => Promise.resolve({
                data: {
                    title: 'title',
                    description: 'desc',
                    people: [],
                    bills: []
                }
            }));
            Vue.prototype.$http = {get};

            const eventForm = mount(EventForm, {propsData: {eventId: 1, buttons: []}});

            await Vue.nextTick();

            expect(get).toHaveBeenCalledWith('/events/1');
            expect(eventForm.vm.title).toBe('title');
            expect(eventForm.vm.description).toBe('desc');
            expect(eventForm.vm.people).toEqual([]);
            expect(eventForm.vm.bills).toEqual([]);
        });

        it('should get user contacts and save them into allPeople', async () => {
            userStore.getUser = jest.fn(() => Promise.resolve({id: 1, email: 'user@email.com'}));
            const contacts = [{id: 2, email: 'notuser@email.com'}];
            const get = jest.fn(() => Promise.resolve({data: contacts}));
            Vue.prototype.$http = {get};

            const eventForm = mount(EventForm, {propsData: {buttons: []}});

            await Vue.nextTick();
            expect(get).toHaveBeenCalledWith('/contact/id/1');
            await Vue.nextTick();
            expect(eventForm.vm.allPeople).toEqual(contacts)
        })
    });

    it('should render', () => {
        const eventForm = mount(EventForm);

        expect(eventForm.html()).toMatchSnapshot();
    })
});
