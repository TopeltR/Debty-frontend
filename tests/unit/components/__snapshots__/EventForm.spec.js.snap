// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm should render 1`] = `
<div>
  <div class="container-fluid">
    <div class="row">
      <nav class="navbar fixed-top w-100 bg-light-gray navbar-light navbar-expand-sm"><button type="button" aria-label="Toggle navigation" aria-controls="nav_collapse" aria-expanded="false" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
        <div class="navbar-brand limegreen brand">debty</div>
        <div id="nav_collapse" class="navbar-collapse collapse" style="display: none;">
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/home" class="nav-link">
                Home
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/events" class="nav-link">
                Events
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/debts" class="nav-link">
                Debts
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/contacts" class="nav-link">
                Contacts
                <!----></a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/profile" class="nav-link">
                Profile
              </a></li>
          </ul>
          <ul class="navbar-nav ml-auto"><span> Welcome, <b></b></span>
            <font-awesome-icon icon="sign-out-alt" class="icon ml-4 mr-2 mt-1"></font-awesome-icon> <span>Sign out</span></ul>
        </div>
      </nav>
    </div>
  </div>
  <div class="diagonal-background container-fluid">
    <div class="container">
      <div class="row">
        <div class="TM30px col-sm-12 col-md-6 offset-md-3">
          <div class="row">
            <h2>Create event</h2>
            <font-awesome-icon id="info" icon="info-circle" class="ml-1"></font-awesome-icon>
            <div aria-hidden="true" class="d-none" style="display: none;">
              <div></div>
            </div>
          </div>
          <div class="row">
            <form class="wide mt-3">
              <div class="form-group"><label for="title">Title:</label> <input type="text" id="title" required="required" placeholder="Enter event title" class="form-control"></div>
              <div class="form-group"><label for="description">Description:</label> <textarea id="description" type="text" placeholder="Enter event description" required="required" class="form-control"></textarea></div>
              <div class="form-group">
                <!---->
                <ul></ul> <label for="user">Add people:</label>
                <div class="row">
                  <div class="PR0px col-9">
                    <div class="autocomplete" id="user" value="[object Object]"><input type="text" placeholder="Name" class="form-control">
                      <ul id="autocomplete-results" class="autocomplete-results" style="display: none;"></ul>
                    </div>
                  </div>
                  <div class="col-3"><button type="button" class="btn btn-primary wide">
                      Add
                    </button></div>
                </div>
              </div>
              <div class="row mt-4">
                <div class="col-12"><button type="button" class="btn wide btn-primary">
                    Create event
                  </button></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
